Improve performance of output cache

mgravell opened the pull request:
# Improve performance of output cache

Reduce allocations and CPU impact of output cache by making changes to the internal implementation (numbers and evidence at bottom)

## Description

The output cache feature is complex; to understand the changes, first we need to discuss the implementation:

- the middleware adds a `OutputCacheContext` / feature
- this attempts to fetch an appropriate `OutputCacheEntry` to use for the existing payload
  - normally, it gets this via a `byte[]` returned from the `IOutputCacheStore`, which it loads into a `MemoryStream` with `BinaryReader`, and manually deserializes into a `FormatterEntry` (parsing the headers into strings, etc - note that the original segments are reconstructed into new `byte[]` copies of the data)
  - `FormatterEntry` has a headers `Dictionary<string, string?[]>`, a `string[]` tags , and a list of `byte[]` buffers
  - it then rewrites this into an `OutputCacheEntry` which has a `HeaderDictionary?`, the same `string[]`, and a `CachedResponseBody` (which wraps a `List<byte[]>`)
- if it is unable to get a matching entry, the output stream is shimmed via an `OutputCacheStream` which writes to the underlying stream but also duplicates the data to a `SegmentWriteStream` (which wraps a `List<byte[]>`) and uses this to write a new `OutputCacheEntry`
  - it may then need to serialize this, so if forms a `FormatterEntry` from the composed `OutputCacheEntry`, writes this via `MemoryStream` and `BinaryWriter` to a `byte[]`, which it passes to the `IOutputCacheStore`

Key observations:

- most of the types cited are `internal`; we have a lot of scope for change
- all of the `byte[]` mentioned are dropped on the floor
- the only time we actually need a right-sized `byte[]` is fore the `IOutputCacheStore` API; this will be addressed separately
- `FormatterEntry` and `OutputCacheEntry` are basically the same thing
- lots of intermediate wrapper objects - `List<T>`, `BinaryReader`, `BinaryWriter`, `MemoryStream`
- the `List<T>` and `MemoryStream` use array-doubling internally, so may involve multiple array allocations behind the scenes
- `BinaryReader` / `BinaryWriter` may use their own internal buffers
- the multiple dictionaries and arrays for headers are redundant; this is pure storage - we can use simpler constructs

Work items:

## merge `FormatterEntry` and `OutputCacheEntry` into a simpler store

we throw away `FormatterEntry`, and simplify `OutputCacheEntry`; we throw away the tags completely (they are not needed in the storage model), use `ReadOnlyMemory<(string, StringValues)>` for the headers (allowing us to use oversized leased arrays), and `ReadOnlySequence<byte>` for the body payload (deleting `CachedResponseBody`)

for the very few places where headers on the `OutputCacheEntry` are inspected, we add a `TryFindHeader` API; this is *incredibly* rare, so a few O(N) scans will be much more efficient than paying the same O(N) cost to *build* a dictionary (plus additional storage), for only a few O(1) fetches

## introduce a custom recyclable sequence API

the new `RecyclableReadOnlySequenceSegment : ReadOnlySequenceSegment<byte>` allows `ReadOnlySequence<byte>` to be constructed cheaply, using two definitions of recycling:

- a small number of segments is kept via a pool, allowing chain details to be recycled
- optionally, recycling a chain may also attempt to detect buffers that can be recycled

## implement a v2 serialization format

fortunately the existing format included version preamble, so we can add a v2 effortlessly; key differences in v2:

1. instead of storing segment subdata, just store the payload length, and write a single contiguous body payload
2. test header name/values against a hard-coded set of known values; if found, store an integer key rather than the length-prefixed UTF8 data

## implement custom reader/writer for the serializer

specifically, we introduce new `ref struct` reader/writer that are wire-compatible with `BinaryReader`/`BinaryWriter`, but minimal; the reader works directly on the `byte[]` returned from `IOutputCacheStore` (later we may want to make this work against `ReadOnlySequence<byte>` for a revised API, but we can defer this), using oversized leased arrays for the headers; if parsing pre-existing v1 data, the body uses the leased segment chain as mentioned above but pointing directly at the original `byte[]` data (which is never reused, so: safe); for v2 data, the body uses a single-segment (no chain) sequence of the payload bytes

the writer targets an `IBufferWriter<byte>`, and in particular a new `RecyclableArrayBufferWriter<byte>` which uses leased arrays (using the single-array model as per `ArrayBufferWriter<byte>`); we still need a `byte[] ToArray()` API to satisfy the `IOutputCacheStore` demands (we can address that later)

## change the output capture to use sequence chains

we change `OutputCacheStream` to use a new `RecyclableSequenceBuilder` instead of `SegmentWriteStream` (which was in "shared", and is now removed), which has similar `Write` APIs, but no longer pretends to be a `Stream`, and uses `RecyclableReadOnlySequenceSegment` and leased buffers for the backing store; the `DetachAndReset()` API hands back a `ReadOnlySequence<byte>` of the captured payload

## recycle the cache-entry as needed

in the middleware, we use a new `ReleaseCachedResponse()` API that cleans up and recycles (if necessary) the `OutputCacheEntry`; this works similarly regardless of whether the current request built the data, or whether it was materialized; to help with this, we decorate the cached response as nullable to formalize that the state could be empty, and fixup call paths

intentionally, we only do this in "known good" code paths - we don't use `finally` / `using` to avoid timing problems with `async` exceptions (which could still have sight of the leased buffers)

## add new performance profiling project

(and backport the same to the original code)

[before](https://github.com/dotnet/aspnetcore/blob/marc/ocbin_pre/src/Middleware/OutputCaching/perf/BenchmarkDotNet.Artifacts/results/Microsoft.AspNetCore.OutputCaching.Benchmark.EndToEndBenchmarks-report-github.md) vs [after](https://github.com/dotnet/aspnetcore/blob/marc/ocbin/src/Middleware/OutputCaching/perf/BenchmarkDotNet.Artifacts/results/Microsoft.AspNetCore.OutputCaching.Benchmark.EndToEndBenchmarks-report-github.md)

key observations:

- write path before could allocate ~5x the payload size; now we allocate ~1x, which is explained by the need of the `byte[]` for the `IOutputCacheStore` API
- read path before allocated ~1x the payload size; now no lost overheads
- performance is improved in all scenarios, but especially on the read path which is now O(1) thanks to (at least for the in-process scenario) using zero-copy for the main payload

gfoidl commented on src/Middleware/OutputCaching/src/RecyclableSequenceBuilder.cs:
```diff
@@ -0,0 +1,127 @@
+    {
+        if (segmentSize <= 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(segmentSize), segmentSize, $"{nameof(segmentSize)} must be greater than 0.");
```
Use the new throw-helpers (ThrowIfNegative)?

mgravell replied:
Couldn't see a ThrowIfNegativeOrZero - I guess I could subtract one...

JamesNK replied:
https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception.throwiflessthanorequal?view=net-8.0

mgravell replied:
heh, there literally is a `ThrowIfNegativeOrZero` - must have missed that (or a multi-TFM thing)

adityamandaleeka commented:
cc @sebastienros

sebastienros commented:
Let's build a benchmark ;)

BrennanConroy reviewed (commented):
Couple high level nits, haven't looked much at the actual code yet

mgravell commented on src/Middleware/OutputCaching/src/Global.cs:
```diff
@@ -0,0 +1,6 @@
+
+using System.Runtime.CompilerServices;
+
+[module: SkipLocalsInit] // all stackalloc usage checked
```
Given C#'s definite assignment, there's only 3 main ways of doing that:

- stackalloc
- calling `out` methods that ignore the semantics (such as `Unsafe.SkipInit')
- intentional pointer math

The last two are fully self-inflicted. The performance benefits of skipping locals init can apply generally, big are observed in particular in stackalloc. But: huge chunks of the runtime now use this mode by default. I genuinely don't think we should shy away from it.

However! I'm OK with compromise - how would you feel about moving the attribute to the method where I'm using the stackalloc for now, and we can look at the rest later on a separate ticket?

I can also run the perf suite in both modes, so we can see the impact in concrete terms.

JamesNK replied:
> However! I'm OK with compromise - how would you feel about moving the attribute to the method where I'm using the stackalloc for now, and we can look at the rest later on a separate ticket?

That's fine. I had a PR that enabled SkipLocalsInit across many projects. Other people were gun-shy about globally enabling it so it was reduced to methods where stackalloc is used.

JamesNK replied:
https://github.com/search?q=repo%3Adotnet%2Faspnetcore%20SkipLocalsInit&type=code

mgravell replied:
I genuinely think we can be more bold about it; IIRC it is on by default for "runtime" ([citation](https://github.com/dotnet/runtime/blob/408729ca215ef5e9ff5127d878b042c4ff3849e9/src/libraries/Directory.Build.targets#L200)), and I could only find a single project that opts out (although there is definitely some validation work per project on the edge cases I mentioned). However, I'm happy to take that as a separate follow up item where we can focus on it as a single dedicated thing, with profiling etc; maybe talk to a few folks like @stephentoub / @jaredpar. For now I'll move it local to just the stackalloc usage.

JamesNK replied:
I made the same points. Closed PR with discussion: https://github.com/dotnet/aspnetcore/pull/26598

I didn't feel strongly enough about it to keep pushing, but you're welcome to develop a good argument 😄 A notable difference in RPS is the best evidence.

mgravell replied:
Useful context. I'm happy to take this as a personal mission, but: I don't need to do that in this PR. I will happily eat the work to get the evidence :)

stephentoub replied:
> It's been 3 years since runtime enabled skip locals init

FWIW, it's been more than that. We switched to using SkipLocalsInit when it was a supported language feature, but before that we achieved the same thing via a linker plugin that stripped out the locals init. I believe we enabled that in NET Core 2.1.

mgravell commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,195 @@
+    {
+        ArgumentOutOfRangeException.ThrowIfNegative(bytes);
+
+        if (offset + bytes > length)
```
Thanks. Not auto-merging now, as I'll do that when I'm at a PC to fix the rest

mgravell commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -131,97 +99,133 @@ private static void Serialize(Stream output, FormatterEntry entry)
+                const int MAX_STACK_BYTES = 256;
+                byte[]? leased = null;
+
+                Span<byte> buffer = bytes <= MAX_STACK_BYTES ? stackalloc byte[bytes] : new(leased = ArrayPool<byte>.Shared.Rent(bytes), 0, bytes);
```
I'll need to check on PC (mobile UI is terrible for this), but this will result in a different sized span, so might need to slice

mgravell commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -131,97 +99,133 @@ private static void Serialize(Stream output, FormatterEntry entry)
+                const int MAX_STACK_BYTES = 256;
+                byte[]? leased = null;
+
+                Span<byte> buffer = bytes <= MAX_STACK_BYTES ? stackalloc byte[bytes] : new(leased = ArrayPool<byte>.Shared.Rent(bytes), 0, bytes);
+                int actual = Encoding.UTF8.GetBytes(value, buffer);
+                Debug.Assert(actual == bytes);
+                writer.WriteRaw(buffer);
```
Heh, see my previous comment (I think we're in agreement here)

danmoseley commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -232,71 +236,262 @@ private static void Serialize(Stream output, FormatterEntry entry)
+    {
+        var arr = CommonHeaders;
+        var pairs = new List<KeyValuePair<string, int>>(arr.Length);
+        for (int i = 0; i < arr.Length; i++)
```
Is that style in this repo (I have no preference)

gfoidl replied:
Yes, `var` is used here in aspnetcore-repo.

mgravell commented:
@gfoidl @BrennanConroy nits resolved - great feedback, thanks!

@sebastienros totally agree we should add a crank profile that stresses output-cache (ideally with both small and large payloads - maybe query-string length?), and get that merged *before* this (so we can compare with/without) - I'll take a stab at that, but I might be looking for some of your input there

BrennanConroy commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -232,71 +226,262 @@ private static void Serialize(Stream output, FormatterEntry entry)
-        return result;
+    static readonly string[] CommonHeaders = new string[]
+    {
+        // to remove values, use ""; DO NOT just remove the line, and do not re-order
```
I think there needs to be a different story for how to remove a value.

If you removed one now (and added `""`) it would mean previously stored values would be read and end up with an empty header value, so we'd lose the value when returning a response.

BrennanConroy commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -232,71 +226,262 @@ private static void Serialize(Stream output, FormatterEntry entry)
 
-        result.Tags = tags;
-        return result;
+    static readonly string[] CommonHeaders = new string[]
```
@Tratcher probably has opinions about this

BrennanConroy commented on src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs:
```diff
@@ -423,15 +402,7 @@ internal async ValueTask FinalizeCacheBodyAsync(OutputCacheContext context)
-                }
-
-                context.CachedResponse.Body = cachedResponseBody;
+                context.CachedResponse.SetBody(cachedResponseBody, recycleBuffers: true); // transfer lifetime from the buffer to the cached response
```
nit: comments not on the same line as the code

BrennanConroy commented on src/Middleware/OutputCaching/test/OutputCacheEntryFormatterTests.cs:
```diff
@@ -1,29 +1,25 @@
-            Headers = new HeaderDictionary(),
-            Tags = Array.Empty<string>()
-        };
+        using var entry = new OutputCacheEntry(KnownTime, 200);
```
```suggestion
        using var entry = new OutputCacheEntry(KnownTime, StatusCodes.Status200OK);
```

BrennanConroy commented on src/Middleware/OutputCaching/test/OutputCacheEntryFormatterTests.cs:
```diff
@@ -38,80 +34,157 @@ public async Task StoreAndGet_StoresAllValues()
+            OutputCacheEntryFormatter.Deserialize(FromHex(KnownV2Payload))
+        );
+    }
+    static byte[] FromHex(string hex)
```
Can replace `Nibble` with `Convert.FromHexString(s);`

BrennanConroy reviewed (commented):
The removal of "Tags" is unclear to me.

adityamandaleeka commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,197 @@
+internal ref struct FormatterBinaryReader
+{
+    // this is effectively a cut-down re-implementation of BinaryReader
+    // from https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs
```
Nit: For ease of future archaeology, link to the commit where it was taken from rather than main.

adityamandaleeka commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,197 @@
+
+    public byte ReadByte()
+    {
+        if (offset >= length)
```
```suggestion
        if (IsEOF)
```

adityamandaleeka commented on src/Middleware/OutputCaching/src/FormatterBinaryWriter.cs:
```diff
@@ -0,0 +1,207 @@
+internal ref struct FormatterBinaryWriter
+{
+    // this is effectively a cut-down re-implementation of BinaryWriter
+    // from https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs
```
Same here, would be better to link to the specific commit.

adityamandaleeka commented on src/Middleware/OutputCaching/src/RecyclableSequenceBuilder.cs:
```diff
@@ -0,0 +1,124 @@
+            else
+            {
+                var toWrite = Math.Min(buffer.Length, available.Length);
+                if (toWrite <= 0)
```
Is this needed? Can it be negative?

mgravell replied:
GetBuffer could in theory return an empty buffer; it isn't meant to, but I don't want to loop forever due to a bad impl. It shouldn't be negative, granted

mgravell commented on src/Middleware/OutputCaching/src/OutputCacheEntry.cs:
```diff
@@ -1,34 +1,143 @@
+    {
+        if (!TryFindHeader(HeaderNames.TransferEncoding, out _))
+        {
+            headers.ContentLength = Body.Length;
```
In terms of? They don't get a content-length header (as per this line) and remain chunked (even though we know the length). This is not a change from the pre-existing impl.

BrennanConroy replied:
Nvm, I just forgot that transfer encoding is where chunked is set.

mgravell commented:
> The removal of "Tags" is unclear to me.

@BrennanConroy from the entry? It is never used, and isn't part of the public API, yet by existing it requires deserialization into `string` each time, so that's allocations and UTF8 decode (plus an allocation/lease for the collection to put them in), every time, for absolutely no reason

martincostello commented on src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:
```diff
@@ -47,6 +45,17 @@ public static async ValueTask StoreAsync(string key, OutputCacheEntry value, Has
 
         string[] tagsArr = tags is { Count: > 0 } ? tags.ToArray() : Array.Empty<string>();
         await store.SetAsync(key, buffer.ToArray(), tagsArr, duration, cancellationToken);
```
I guess you meant to delete this line as part of the merge?

mgravell committed:
optimize Output Cache; no API changes yet - all internal:
- unify OutputCacheEntry and FormatterEntry
- leased buffers for headers, tags, etc (dispose on way out)
- use ReadOnlySequence<byte> instead of List<byte[]> with recyclable segments
- avoid copying the payload data once fectched
- serialization tweak: use common headers (not yet listed)

finish porting tests; all good

migrate writes to IBufferWriter<byte>, using recyclable array buffer writer (similar to MemoryStream, but: faster)

use pooled buffers when buffering output-cache payloads RROSS should be able to recycle buffers as needed

implement header name/value lookup buffer

add bench project

note memory overhead in bench

full bench suite

add benchmark result

- don't store tags in the cache payload
- don't store segment details in the cache payload

whitespace

include test that uses body-writer rather than stream

tidy up benchmarks

remove pipe impl; we don't need it (proven in bench)

add a few more known headers; don't store request-id

don't write empty segments

fix buffer cleanup paths

revert changes to shared SegmentWriteStream

update numbers

use fixed size stackalloc

Update src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

Co-authored-by: Günther Foidl <gue@korporal.at>

- remove global SkipLocalsInit
- add accessibility on all new members
- in length checks, use - over + to avoid overflow

nits

relocate perf to microbenchmarks

use standard project structure for micro-benchmarks

use "actual" instead of "bytes" to avoid a "pop"/"ld" in the "release" case

avoid a memcopy by writing directly to the buffer bytes (and increasing DRY)

Update src/Middleware/OutputCaching/src/OutputCacheEntry.cs

Co-authored-by: Brennan <brecon@microsoft.com>

Update src/Middleware/OutputCaching/test/OutputCacheEntryFormatterTests.cs

Co-authored-by: Brennan <brecon@microsoft.com>

fix PR nits from #48450

fix sln (dead project)

Update src/Middleware/OutputCaching/src/OutputCacheEntry.cs

Co-authored-by: Brennan <brecon@microsoft.com>

DRY nit

moar nits

use leased buffer for tags when calling SetAsync

Update src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

Co-authored-by: Brennan <brecon@microsoft.com>

use [LoggerMessage] for logging

merge submodule delete

bump

optimize Output Cache; no API changes yet - all internal:
- unify OutputCacheEntry and FormatterEntry
- leased buffers for headers, tags, etc (dispose on way out)
- use ReadOnlySequence<byte> instead of List<byte[]> with recyclable segments
- avoid copying the payload data once fectched
- serialization tweak: use common headers (not yet listed)

sln fix

dammit

mgravell commented:
(rebased on top of redis output cache changes)

ReubenBond commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,197 @@
+
+        result |= (uint)byteReadJustNow << (MaxBytesWithoutOverflow * 7);
+        return (int)result;
+    }
```
I believe I've shared it before, but here are some virtually branch-free variants of these read/write u32/u64 methods: https://github.com/dotnet/orleans/blob/964ed094a5f77b3181c330a0cf33c58e19b48b85/src/Orleans.Serialization/Buffers/Reader.cs#L879

mgravell replied:
the "using a different format" is part of the problem; we want to retain compat, since (at least in theory) we could have pre-existing data; in reality, this is trivial serialization - we aren't serializing deeply nested object graphs with "parent / child" repetition, for example: I don't think we'll see any genuine benefit from hyper-focusing on this, but we *will* add risk and complexity

ReubenBond replied:
Similar to the other comment, I'm ok with not optimizing this so much since it's not general-purpose, but it's worth pointing out regardless.

ReubenBond commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,197 @@
+        result |= (uint)byteReadJustNow << (MaxBytesWithoutOverflow * 7);
+        return (int)result;
+    }
+
```
If we assume that smaller magnitude numbers are more common than larger magnitude numbers, then zig-zag encoding is more efficient for signed integers. Casting works but it results in 5-byte encodings for the likely common `-1` case and 10-byte encodings for a -1 64-bit int. https://github.com/dotnet/orleans/blob/964ed094a5f77b3181c330a0cf33c58e19b48b85/src/Orleans.Serialization/Utilities/VarIntWriterExtensions.cs#L50

mgravell replied:
in theory yes they are (it is a general purpose API), but: the heart of this needs to support pre-existing data written via `BinaryWriter`, so: we can't change the layout at least for the old usage; we could say "v2 always uses zigzag"; however, if we look at the callers it turns out that everything *is* positive, except perhaps for offset minutes which is more mercurial; in reality, I don't think we need to worry about this nuance; we *could* add unsigned methods, but... honestly, I think it is fine as it is

ReubenBond replied:
It's not optimal but it's ok since this is not a general-purpose serializer

ReubenBond commented on src/Middleware/OutputCaching/src/IOutputCacheBufferStore.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <param name="destination">The location to which the value should be written.</param>
+    /// <param name="cancellationToken">Indicates that the operation should be cancelled.</param>
+    /// <returns><c>True</c> if the response cache entry if it exists; otherwise <c>False</c>.</returns>
+    ValueTask<bool> TryGetAsync(string key, PipeWriter destination, CancellationToken cancellationToken);
```
Am I right in assuming we're using `PipeWriter` instead of `IBufferWriter<byte>` so that responses can be streamed to the caller asynchronously (with backpressure)?

mgravell replied:
Yes, that's the intent - that implementations can respect back-pressure semantics

captainsafia commented on src/Middleware/OutputCaching/perf/Microbenchmarks/Microsoft.AspNetCore.OutputCaching.Microbenchmarks.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <Reference Include="Microsoft.AspNetCore.OutputCaching" />
+    <Reference Include="BenchmarkDotNet" />
+    <Reference Remove="Microsoft.CodeAnalysis.PublicApiAnalyzers" />
```
> investigating the overheads;

Can you clarify what you mean here? I would think that it's not necessary to include the PublicApi analyzers on a benchmark app where we are not shipping an API.

mgravell commented on src/Middleware/OutputCaching/src/CacheEntryHelpers.cs:
```diff
@@ -20,19 +20,13 @@ internal static long EstimateCachedResponseSize(OutputCacheEntry cachedResponse)
 
             // Headers
-            if (cachedResponse.Headers != null)
+            foreach (var item in cachedResponse.Headers.Span)
```
@captainsafia because it is now a `ReadOnlyMemory<T>` rather than a `HeaderDictionary?`

mgravell commented on src/Middleware/OutputCaching/src/CacheEntryHelpers.cs:
```diff
@@ -20,19 +20,13 @@ internal static long EstimateCachedResponseSize(OutputCacheEntry cachedResponse)
-            {
-                size += cachedResponse.Body.Length;
-            }
+            size += cachedResponse.Body.Length;
```
@captainsafia it is now a `ReadOnlySequence<byte>` - worst case is that it is empty, which doesn't need any extra work

mgravell committed:
use `_field` naming in `FormatterBinaryReader.cs`

mgravell commented on src/Middleware/OutputCaching/src/FormatterBinaryReader.cs:
```diff
@@ -0,0 +1,197 @@
+    // additionally, we add support for reading a string with length specified by the caller (rather than handled automatically),
+    // and in-place (zero-copy) BLOB reads
+
+    private readonly ReadOnlyMemory<byte> original; // used to allow us to zero-copy chunks out of the payload
```
applied

mgravell committed:
move log methods to dedicated file

mgravell commented on src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs:
```diff
@@ -110,8 +108,12 @@ private async Task InvokeAwaited(HttpContext httpContext, IReadOnlyList<IOutputC
                 {
-                    if (await TryServeFromCacheAsync(context, policies))
+                    bool served = await TryServeFromCacheAsync(context, policies);
+                    context.ReleaseCachedResponse(); // release even if not served due to failing conditions
```
the lifetime is a bit more complex than that - the cached response is changed in a few spots; in all the `return` cases you may be right, but the one you highlight here: can get overstomped again later, validly; I was also trying to not recycle buffers in any "cancellation token abort" scenarios, as I don't want to recycle anything if there's even a remote chance that async code (the output cache implementation) is still touching that buffer - I'd rather drop them on the floor than get competition

that was my logic; happy to add some comments, and maybe a `bool releaseResponse` for the common `return` case (so the only thing repeated is `releaseResponse = true` ?

BrennanConroy replied:
Maybe I'm missing something, but it looks like every single spot this method exits from, the cached response is released. And if I am missing something and it's more complex than that, there should probably be comments explaining it.

> I don't want to recycle anything if there's even a remote chance that async code (the output cache implementation) is still touching that buffer

You can preserve that behavior:
```csharp
var hasException = false;
try
{
}
catch
{
    hasException = true;
    throw;
}
finally
{
    if (!hasException)
    {
        // release
    }
}
```

mgravell replied:
applied

BrennanConroy commented on src/Middleware/OutputCaching/src/FormatterBinaryWriter.cs:
```diff
@@ -0,0 +1,207 @@
+    private void RequestNewBuffer()
+    {
+        Flush();
+        var span = target.GetSpan(1024);
```
> ultimately any implementation can ignore us

Well, ignore the lower bound, they should always give a buffer `>=` the value passed in.

Sure, a comment is fine. It'd be interesting to see if real world examples could benefit from a larger buffer. 😃

mgravell replied:
applied

mgravell committed:
simplify cached response cleanup

mgravell committed:
add comment re buffer size
